core/src/env/srml/calls.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of ink!.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::env::EnvTypes;
+use parity_codec::{
+    Decode,
+    Encode,
+    Input,
+    Output,
+};
+use core::convert::TryInto;
+
+#[cfg_attr(feature = "test-env", derive(Debug, Clone, PartialEq, Eq))]
+pub enum Address<T: EnvTypes> {
+    Id(T::AccountId),
+    Index(T::AccountIndex),
+}
+
+fn need_more_than<T: PartialOrd>(a: T, b: T) -> Option<T> {
+    if a < b { Some(b) } else { None }
+}
+
+/// Decode impl copied from
+/// https://github.com/paritytech/substrate/blob/ec62d24c602912f07bbc416711376d9b8e5782c5/srml/indices/src/address.rs#L61
```

Comment:
Please make use of markdown and properly name the link, a la:

```markdown
Decode implementation copied over from Substrate `Address` that can be found [here](substrate-address).

[substrate-address]: https://github.com/paritytech/substrate/blob/ec62d24c602912f07bbc416711376d9b8e5782c5/srml/indices/src/address.rs#L61
```

Also please notify readers that this implementation must be kept in sync for it to work.
This is a very fragile part of the code base.

Reply:
Suggestion implemented in https://github.com/paritytech/ink-types-node-runtime/pull/3